src/qt/walletmodel.cpp

```diff
@@ -756,3 +779,296 @@ bool WalletModel::getDefaultWalletRbf() const
+    const std::string toaddress = it->second.toaddress;
+
+    updatedNameNewReturn.ok = true;
+    updatedNameNewReturn.hex = txid;
+    updatedNameNewReturn.rand = rand;
+    updatedNameNewReturn.data = strData;
+    if(!toaddress.empty ())
+        updatedNameNewReturn.toaddress = toaddress;
+
+    it->second = updatedNameNewReturn;
+
+    UniValue uniNameUpdateData(UniValue::VOBJ);
+    uniNameUpdateData.pushKV ("txid", txid);
+    uniNameUpdateData.pushKV ("rand", rand);
+    uniNameUpdateData.pushKV ("data", strData);
+    if(!toaddress.empty ())
+        uniNameUpdateData.pushKV ("toaddress", toaddress);
+
+    std::string jsonData = uniNameUpdateData.write();
+    LogPrintf ("Writing pending name_firstupdate %s => %s\n", strName.c_str(), jsonData.c_str());
+
+    CWalletDBWrapper& dbw = wallet->GetDBHandle();
+    CWalletDB(dbw).WriteNameFirstUpdate(strName, jsonData);
+
+    return tr("");
+}
+
+std::string WalletModel::completePendingNameFirstUpdate(std::string &name, std::string &rand, std::string &txid, std::string &data, std::string &toaddress)
+{
+    JSONRPCRequest jsonRequest;
+    UniValue params(UniValue::VOBJ);
+    UniValue res;
+    std::string errorStr;
+
+    params.push_back (Pair("name", name));
+    params.push_back (Pair("rand", rand));
+    params.push_back (Pair("tx", txid));
+    params.push_back (Pair("value", data));
+    if(!toaddress.empty())
+        params.push_back (Pair("toaddress", toaddress));
+
+    jsonRequest.strMethod = std::string("name_firstupdate");
+    jsonRequest.params = params;
+    jsonRequest.fHelp = false;
+
+    LogPrintf("executing name_firstupdate name=%s rand=%s tx=%s value=%s\n",
+        name.c_str(), rand.c_str(), txid.c_str(), data.c_str());
+
+    try {
+        res = name_firstupdate (jsonRequest);
+    }
+    catch (const UniValue& e) {
+        UniValue message = find_value( e, "message");
+        errorStr = message.get_str();
+        LogPrintf ("name_firstupdate error: %s\n", errorStr.c_str());
+    }
+    return errorStr;
+}
+
+void WalletModel::sendPendingNameFirstUpdates()
+{
+    for (std::map<std::string, NameNewReturn>::iterator i = pendingNameFirstUpdate.begin();
+         i != pendingNameFirstUpdate.end(); )
+    {
+        JSONRPCRequest jsonRequest;
+        UniValue params1(UniValue::VOBJ);
+        UniValue res1, val;
+        // hold the error returned from name_firstupdate, via
+        // completePendingNameFirstUpdate, or empty on success
+        // this will drive the error-handling popup
+        std::string completedResult;
+
+        std::string name = i->first;
+        std::string txid = i->second.hex;
+        std::string rand = i->second.rand;
+        std::string data = i->second.data;
+        std::string toaddress = i->second.toaddress;
+
+        params1.push_back (Pair("txid", txid));
+        jsonRequest.params = params1;
+        // if we're here, the names doesn't exist
+        // should we remove it from the DB?
```

Comment:
What do you mean by this comment?  There has not been any branch yet, so how do we already know anything about the names and DB at this point?